nixos/modules/services/x11/display-managers/sddm.nix

```diff
@@ -69,6 +69,13 @@ let
       Session = autoLoginSessionName;
       Relogin = cfg.autoLogin.relogin;
     };
+  } // lib.optionalAttrs cfg.enableKwinWayland {
+    General = {
+      GreeterEnvironment = "QT_PLUGIN_PATH=${pkgs.plasma5Packages.layer-shell-qt}/${pkgs.plasma5Packages.qtbase.qtPluginPrefix},QT_WAYLAND_SHELL_INTEGRATION=layer-shell";
```

Comment:
`QT_PLUGIN_PATH` here is not applied somehow. Don't know why.

Reply:
Why to force layer-shell when upstream uses fullscreen-shell-v1 (I see it's hardcoded in sddm source code)? I don't think kwin supports layer-shell atm, I heard there were talks about switching to it in Plasma 6, but right now kwin support only Plasma's own analogue, plasma-shell.

Reply:
I just copied the config from upstream Plasma to see if it worked (it didn't), and never got back around to looking into it. I might mess with it more today.

Reply:
If someone is interested in the link: https://invent.kde.org/plasma/plasma-workspace/-/blob/master/sddm-wayland-session/plasma-wayland.conf . But Lt I also never got that working (when the version was not released) due the error I posted below

Reply:
sddm code seem to have some special handling of QT_PLUGIN_PATH passing it through processes, so passing it to systemd service might be a better choice. I also don't understand how this is supposed to work given that it doesn't merge the value with the older one but overwrites it, so Qt likely won't even see the wayland platform itself?

Reply:
The wrapper appends to QT_PLUGIN_PATH so it should be fine.

Reply:
Ah, so sddm-greeter itself is also wrapped...

Reply:
> Why to force layer-shell when upstream uses fullscreen-shell-v1 (I see it's hardcoded in sddm source code)? I don't think kwin supports layer-shell atm, I heard there were talks about switching to it in Plasma 6, but right now AFAIK kwin supports only Plasma's own analogue, plasma-shell.

Sorry I'm not familiar with the shell integration envvar. It is copied from [Arch wiki](https://wiki.archlinux.org/title/SDDM). Just I found that without this envvar, it shows a black screen, but greeter and kwin are both running somehow. Debugging their behaviors outside a regular user session is quite painful. And I'm not really sure if `nixos-rebuild test` can reload them correctly.

Reply:
Never mind, @Shawn8901 says it's a config from KDE as well

Reply:
Is there an easy way to apply this PR (and don't rebuild half of my system due to plasma-framework/kwin change)? I'd like to experiment with SDDM on Wayland, the current SDDM is driving me nuts.

Reply:
You can remove the Plasma change locally and add kirigami directly to SDDM's build inputs instead.

Reply:
Looks like those attrsets aren't really getting merged, that's what I have in those sections of the config file:
```
[General]
DisplayServer=wayland
GreeterEnvironment=QT_PLUGIN_PATH=/nix/store/hirgwasga62g8mc0id2qymcj55m4p9k9-layer-shell-qt-5.27.6/lib/qt-5.15.9/plugins,QT_WAYLAND_SHELL_INTEGRATION=layer-shell
InputMethod=
```
```
[Wayland]
CompositorCommand=/nix/store/4p28alhhhpgyw4aifhy3dq15qz6hwjp1-kwin-5.27.6/bin/kwin_wayland --no-global-shortcuts --no-lockscreen --locale1
```

No EnableHiDPI, no SessionDir, no {Halt,Reboot}Command, no Numlock, no DefaultSession

Reply:
Ok, I think I've found where the variable is being reset. sddm uses Qt's QProcessEnvironment::insert(const [QProcessEnvironment](https://doc.qt.io/qt-5/qprocessenvironment.html#QProcessEnvironment) &e) API shadowing feature a lot (in other words, it inserts a list of variables into another one and if there are conflicts, the newly inserted ones silently get prioritized).

GreeterEnvironment option is being read in [HelperApp.cpp](https://github.com/sddm/sddm/blob/07631f2ef00a52d883d0fd47ff7d1e1a6bc6358f/src/helper/HelperApp.cpp#L170-L187). The array is then being shadowed by the result of pam_getenvlist in [PamBackend::openSession](https://github.com/sddm/sddm/blob/07631f2ef00a52d883d0fd47ff7d1e1a6bc6358f/src/helper/backend/PamBackend.cpp#L276) and by setclassenvironment calls in [Backend::openSession](https://github.com/sddm/sddm/blob/07631f2ef00a52d883d0fd47ff7d1e1a6bc6358f/src/helper/Backend.cpp#L82-L101).

My guess is QT_PLUGIN_PATH most likely gets overriden by pam_getenvlist as /etc/pam/environment contains QT_PLUGIN_PATH thanks to `environment.profileRelativeSessionVariables`.

Reply:
Patching nixpkgs to get rid of QT_PLUGIN_PATH in environment.profileRelativeSessionVariables makes sddm start on my machine. But it obviously doesn't see Plasma Wayland session due to the above-mentioned problem.

Reply:
replacing `sessionEnv.insert(m_pam->getEnv());` with `sessionEnv = m_pam->getEnv().insert(sessionEnv);` should also work in theory

Reply:
Hmm, I wonder if we should take this upstream then.

Reply:
Yeah, that's a good question, depends on how it's intended to work